pkg/common/ipsec/utils.go

```
@@ -15,16 +17,25 @@ const (
 	markStateOut = 0xe00
 )
 
-func CountUniqueIPsecKeys(states []netlink.XfrmState) int {
+func CountUniqueIPsecKeys(states []netlink.XfrmState) (int, error) {
 	keys := make(map[string]bool)
 	for _, s := range states {
-		if s.Aead == nil {
+		if s.Aead != nil {
+			keys[string(s.Aead.Key)] = true
+			continue
+		}
+		if s.Auth == nil && s.Crypt == nil {
 			continue
 		}
-		keys[string(s.Aead.Key)] = true
+		// We don't support XFRM states with only one of auth or crypt set.
+		if s.Auth == nil || s.Crypt == nil {
+			return 0, fmt.Errorf("an unsupported key with only Auth or Crypt set has been found")
```

Should we really error out when we find an invalid entry, or rather just skip in and log the error?

If the former, I guess, we should also error out on `s.Aead != nil && (s.Auth != nil || s.Crypt != nil)`. If we restructure the code as follows, we should catch all invalid cases (including 3 nils):

```
if (s.Aead != nil) {
    ...
    continue
}
if (s.Auth != nil && s.Crypt != nil) {
    ...
    continue
}
return 0, fmt.Errorf(...)
```

---

Yeah, looks like you are right `3 nils` is another invalid situation.

---

I've refactored `CountUniqueIPsecKeys` with your suggestion, thanks!

Regarding error, I've checked that it won't impact `cilium metrics list` if IPSec is not enabled.
I don't have a strong opinion here but I think error will make it more visible, isn't it?

---

On second thought, my suggestion is not completely correct, it still misses `s.Aead != nil && (s.Auth != nil || s.Crypt != nil)`. This one should work better and catch all invalid combinations (by explicitly allowing only valid ones):

```
if (s.Aead != nil && s.Auth == nil && s.Crypt == nil) {
    ...
    continue
}
if (s.Aead == nil && s.Auth != nil && s.Crypt != nil) {
    ...
    continue
}
return 0, fmt.Errorf(...)
```

> I think error will make it more visible, isn't it?

Yeah, let's keep the error as long as it doesn't break anything. Let's just make this text `"an unsupported key with only Auth or Crypt set has been found"` more generic, it could be other invalid combinations.

---

Fixed.